src/lp_pricer/lp_pricer.cairo

```diff
@@ -0,0 +1,391 @@
+
+        // [Effect] Set owner
+        let mut state: Ownable::ContractState = Ownable::unsafe_new_contract_state();
+        Ownable::InternalImpl::initializer(ref state, admin_address);
+
+        // [Effect] Set Oracle
+        let oracle = IOracleABIDispatcher { contract_address: oracle_address };
+        self.oracle.write(oracle);
+    }
+
+    // ================== PUBLIC ABI ==================
+
+    #[external(v0)]
+    impl LpPricerImpl of ILpPricer<ContractState> {
+        /// Prices a pool in USD.
+        /// 
+        /// The current formula is:
+        /// (token A reserve * token A price + token B reserve * token B price) / total supply
+        /// 
+        /// The prices of the underlying tokens are retrieved from the Pragma Oracle.
+        fn get_pool_usd_price(self: @ContractState, pool_address: ContractAddress) -> u256 {
+            // [Check] Pool is supported
+            assert(self.is_supported_pool(pool_address), errors::UNSUPPORTED_POOL);
+
+            // [Effect] Get the addresses, symbols & reserves 
+            let pool = self.supported_pools.read(pool_address);
+            let (token_a_address, token_b_address) = pool.get_tokens_addresses();
+            let (token_a_id, token_b_id) = pool.get_tokens_symbols();
+            let (token_a_reserve, token_b_reserve) = pool.dispatcher.get_reserves();
+
+            // [Effect] Get token prices
+            let oracle = self.oracle.read();
+            let token_a_price = get_currency_price_in_usd(oracle, token_a_id);
+            let token_b_price = get_currency_price_in_usd(oracle, token_b_id);
+
+            (token_a_reserve * token_a_price + token_b_reserve * token_b_price) / pool.total_supply
+        }
+
+        /// Register a pool into the supported list.
+        /// 
+        /// Can only be called by the admin.
+        fn add_pool(ref self: ContractState, pool_address: ContractAddress) {
+            // [Check] Only admin
+            assert_only_admin();
+            // [Check] Pool is not already registered
+            assert(!self.is_supported_pool(pool_address), errors::POOL_ALREADY_REGISTED);
+
+            // [Effect] Fetch the underlying token of the pool (address + symbol)
+            let pool_dispatcher = IPoolDispatcher { contract_address: pool_address };
+            let (token_a_address, token_b_address) = fetch_tokens_addresses(pool_dispatcher);
+            let (token_a_id, token_b_id) = fetch_tokens_symbols(token_a_address, token_b_address);
+
+            // [Check] Assert that both pool assets are supported by Pragma
+            let oracle = self.oracle.read();
+            assert(currency_is_supported(oracle, token_a_id), errors::UNSUPPORTED_CURRENCY);
+            assert(currency_is_supported(oracle, token_b_id), errors::UNSUPPORTED_CURRENCY);
+
+            // [Effect] Add the pool to the storage
+            let pool = Pool {
+                id: pool_dispatcher.symbol(),
```

Comment:
I think before adding the pool to the storage, we might want to make sure that the id is not null, for safety, since other mecanisms rely on this id